File: tests/fixtures/cdk_test_project/vulnerable_stack.ts
import * as cdk from "aws-cdk-lib";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as rds from "aws-cdk-lib/aws-rds";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as iam from "aws-cdk-lib/aws-iam";
import { Construct } from "constructs";

export class VulnerableStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const publicBucket = new s3.Bucket(this, "PublicBucket", {
      publicReadAccess: true,
      versioned: false,
    });

    const unprotectedBucket = new s3.Bucket(this, "UnprotectedBucket", {
      versioned: true,
    });

    const unencryptedDb = new rds.DatabaseInstance(this, "UnencryptedDB", {
      engine: rds.DatabaseInstanceEngine.POSTGRES,
      storageEncrypted: false,
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.BURSTABLE3,
        ec2.InstanceSize.MICRO,
      ),
      vpc: undefined,
    });

    const unencryptedVolume = new ec2.Volume(this, "UnencryptedVolume", {
      availabilityZone: "us-east-1a",
      size: cdk.Size.gibibytes(10),
      encrypted: false,
    });

    const unprotectedTable = new dynamodb.Table(this, "UnprotectedTable", {
      partitionKey: {
        name: "id",
        type: dynamodb.AttributeType.STRING,
      },
    });

    const wildcardPolicy = new iam.PolicyStatement({
      actions: ["*"],
      resources: ["arn:aws:s3:::my-bucket/*"],
    });

    const wildcardResourcePolicy = new iam.PolicyStatement({
      actions: ["s3:GetObject", "s3:PutObject"],
      resources: ["*"],
    });

    const adminRole = new iam.Role(this, "AdminRole", {
      assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess"),
      ],
    });

    const openSgIpv4 = new ec2.SecurityGroup(this, "OpenSecurityGroupIPv4", {
      vpc: undefined,
      allowAllOutbound: true,
    });
    openSgIpv4.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(22),
      "SSH from anywhere",
    );

    const openSgIpv6 = new ec2.SecurityGroup(this, "OpenSecurityGroupIPv6", {
      vpc: undefined,
      allowAllOutbound: false,
    });
    openSgIpv6.addIngressRule(
      ec2.Peer.anyIpv6(),
      ec2.Port.tcp(443),
      "HTTPS from anywhere IPv6",
    );
  }
}

File: tests/fixtures/cdk_test_project/vulnerable_stack.py
"""Test CDK stack with security vulnerabilities for extraction verification."""

from aws_cdk import Stack
from aws_cdk import aws_dynamodb as dynamodb
from aws_cdk import aws_ec2 as ec2
from aws_cdk import aws_iam as iam
from aws_cdk import aws_rds as rds
from aws_cdk import aws_s3 as s3
from constructs import Construct


class VulnerableStack(Stack):
    """CDK stack with intentional security misconfigurations for testing.

    Coverage for all CDK security rules:
    - aws_cdk_encryption: RDS, EBS, DynamoDB
    - aws_cdk_iam_wildcards: actions, resources, AdministratorAccess
    - aws_cdk_s3_public: public_read_access, missing block_public_access
    - aws_cdk_security_groups: 0.0.0.0/0, ::/0, allow_all_outbound
    """

    def __init__(self, scope: Construct, construct_id: str, **kwargs):
        super().__init__(scope, construct_id, **kwargs)

        s3.Bucket(self, "PublicBucket", public_read_access=True, versioned=False)

        s3.Bucket(self, "UnprotectedBucket", versioned=True)

        rds.DatabaseInstance(
            self,
            "UnencryptedDB",
            engine=rds.DatabaseInstanceEngine.POSTGRES,
            storage_encrypted=False,
            instance_type=ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE3, ec2.InstanceSize.MICRO),
            vpc=None,
        )

        ec2.Volume(
            self, "UnencryptedVolume", availability_zone="us-east-1a", size=10, encrypted=False
        )

        dynamodb.Table(
            self,
            "UnprotectedTable",
            partition_key=dynamodb.Attribute(name="id", type=dynamodb.AttributeType.STRING),
        )

        iam.PolicyStatement(actions=["*"], resources=["arn:aws:s3:::my-bucket/*"])

        iam.PolicyStatement(actions=["s3:GetObject", "s3:PutObject"], resources=["*"])

        iam.Role(
            self,
            "AdminRole",
            assumed_by=iam.ServicePrincipal("lambda.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name("AdministratorAccess")
            ],
        )

        open_sg_ipv4 = ec2.SecurityGroup(
            self, "OpenSecurityGroupIPv4", vpc=None, allow_all_outbound=True
        )
        open_sg_ipv4.add_ingress_rule(
            peer=ec2.Peer.any_ipv4(), connection=ec2.Port.tcp(22), description="SSH from anywhere"
        )

        open_sg_ipv6 = ec2.SecurityGroup(
            self, "OpenSecurityGroupIPv6", vpc=None, allow_all_outbound=False
        )
        open_sg_ipv6.add_ingress_rule(
            peer=ec2.Peer.any_ipv6(),
            connection=ec2.Port.tcp(443),
            description="HTTPS from anywhere IPv6",
        )
